### ISSUE TITLE
Improved performance

### ISSUE THREAD

**Issue by haberdashPI** [haberdashPI/SignalOperators.jl#5]
I'm slowly working through benchmarks to make sure each `sampleat!` implementation is entirely stack based.

Types to check
- [x] Arrays
- [x] Functions
- [x] Numbers
- [x] `until` and `after`
- [x] `pad`, `append`
- [x] `mapsignal` and derivatives
- [x] `filtersignal` and derivatives

**Comment by haberdashPI** [comment]
This will probably benefit from the refactoring in #2

**Comment by haberdashPI** [comment]
Idea: since, `sink_helper!` dispatches on checkpoint objects we can get efficient code for each channel count by including channel count in the type signature of checkpoint objects.

**Comment by haberdashPI** [comment]
Questions for current [benchmark test](https://github.com/haberdashPI/SignalOperators.jl/blob/master/test/benchmarks.jl)

Why are there so many allocations?
Is @Base.propagate_inbounds working?
Is inline of single loop functions working?
How does the @code_warntype of `sampleat!` look?

**Comment by haberdashPI** [comment]
My earlier idea turns out to be the case: for `sampleat!`, if we return a tuple, we need to know how many channels there are based on the type of the signal.

It is probably not necessary to use a tuple, I think a lazy array type that just artificially broadcasts across the channels should do fine.

**Comment by haberdashPI** [comment]
Notes:

1. The use of views may be an issue; see Julia issue [#14955](https://github.com/JuliaLang/julia/issues/14955)

2. There's something wrong with how AxisArrays are handled, I think. @trace (from [Traceur](http://traceur.junolab.org/latest/)) seems to think there are dynamic dispatches to a number of AxisArrray methods.

**Comment by haberdashPI** [comment]
I've convinced myself that there are likely not any type instabilities in the code well enough that I think it is worth seeing how performance improves if I avoid array views. 

There still could be an issue with AxisArrays, and it would be worth attempting some code that avoids them (currently they are always used) just to see how that goes, if removing the array views doesn't work.

**Comment by haberdashPI** [comment]
There is still a type instability:

```julia
julia> z = mix(x,y) |> tosamplerate(1000Hz)
julia> @trace SignalOperators.sampleat!(out,z,SignalOperators.SignalTrait(z),1,1,checks[5]) modules=[SignalOperators]
┌ Warning: dynamic dispatch to SignalOperators.writesink!(SignalOperators.one_sample, 1, %new(FillArrays.Fill{Float64,1,Tuple{Base.OneTo{Int64}}}, 0.0, Core.tuple(%new(Base.OneTo{Int64}, Base.ifelse(Base.slt_int(Base.arraysize(Base.getfield(Base.getfield(Base.getfield(Base.getfield(x, padded_signals), 1, $(Expr(:boundscheck))), signal), data), 2), 0), 0, Base.arraysize(Base.getfield(Base.getfield(Base.getfield(Base.getfield(x, padded_signals), 1, $(Expr(:boundscheck))), signal), data), 2))))))
└ @ ~/Documents/work/projects/SignalOperators/src/mapsignal.jl:-1
┌ Warning: dynamic dispatch to SignalOperators.writesink!(SignalOperators.one_sample, 1, %new(AxisArrays.AxisArray{Float64,1,SubArray{Float64,1,Array{Float64,2},Tuple{Int64,Base.Slice{Base.OneTo{Int64}}},true},Tuple{AxisArrays.Axis{:channel,UnitRange{Int64}}}}, %new(SubArray{Float64,1,Array{Float64,2},Tuple{Int64,Base.Slice{Base.OneTo{Int64}}},true}, Base.getfield(Base.getfield(Base.getfield(Base.getfield(x, padded_signals), 2, $(Expr(:boundscheck))), signal), data), Core.tuple(j, %new(Base.Slice{Base.OneTo{Int64}}, %new(Base.OneTo{Int64}, Base.ifelse(Base.slt_int(Base.arraysize(Base.getfield(Base.getfield(Base.getfield(Base.getfield(x, padded_signals), 2, $(Expr(:boundscheck))), signal), data), 2), 0), 0, Base.arraysize(Base.getfield(Base.getfield(Base.getfield(Base.getfield(x, padded_signals), 2, $(Expr(:boundscheck))), signal), data), 2))))), Base.sub_int(Base.add_int(Base.add_int(1, Base.mul_int(Base.sub_int(j, 1), 1)), Base.mul_int(0, Base.mul_int(1, Base.sub_int(Base.ifelse(Base.slt_int(Base.arraysize(Base.getfield(Base.getfield(Base.getfield(Base.getfield(x, padded_signals), 2, $(Expr(:boundscheck))), signal), data), 1), 0), 0, Base.arraysize(Base.getfield(Base.getfield(Base.getfield(Base.getfield(x, padded_signals), 2, $(Expr(:boundscheck))), signal), data), 1)), 0)))), Base.mul_int(Base.mul_int(1, Base.sub_int(Base.ifelse(Base.slt_int(Base.arraysize(Base.getfield(Base.getfield(Base.getfield(Base.getfield(x, padded_signals), 2, $(Expr(:boundscheck))), signal), data), 1), 0), 0, Base.arraysize(Base.getfield(Base.getfield(Base.getfield(Base.getfield(x, padded_signals), 2, $(Expr(:boundscheck))), signal), data), 1)), 0)), 1)), Base.mul_int(1, Base.sub_int(Base.ifelse(Base.slt_int(Base.arraysize(Base.getfield(Base.getfield(Base.getfield(Base.getfield(x, padded_signals), 2, $(Expr(:boundscheck))), signal), data), 1), 0), 0, Base.arraysize(Base.getfield(Base.getfield(Base.getfield(Base.getfield(x, padded_signals), 2, $(Expr(:boundscheck))), signal), data), 1)), 0))), Core.tuple(Base.getfield(Base.getfield(Base.getfield(Base.getfield(Base.getfield(x, padded_signals), 2, $(Expr(:boundscheck))), signal), axes), 2, true))))
└ @ ~/Documents/work/projects/SignalOperators/src/mapsignal.jl:-1
2-element Array{Float64,1}:
 0.3300021471853958
 0.8360567643791674
```

**Comment by haberdashPI** [comment]
This prior issues turned out to be because of a number of different sources of memory allocation during `sampleat!`; I've now got a handle on it and am going through each of the types to make sure no allocations are happening for each type of signal.

**Comment by haberdashPI** [comment]
Closed by e0ec3f0ba733d687ec090c0a58ab72240e9a4ea3
